hack/tools/janitor/janitor.go

```diff
@@ -0,0 +1,245 @@
+		}
+		vmsToDelete = append(vmsToDelete, vm)
+	}
+
+	// PowerOff vms which are still running. Triggering PowerOff for a VM results in a task in vSphere.
+	poweroffTasks := []*object.Task{}
+	for _, vm := range vmsToDeleteAndPoweroff {
+		log.Info("Powering off vm in vSphere", "vm", vm.managedObject.Config.Name)
+		if s.dryRun {
+			// Skipping actual PowerOff on dryRun.
+			continue
+		}
+		task, err := vm.object.PowerOff(ctx)
+		if err != nil {
+			return err
+		}
+		poweroffTasks = append(poweroffTasks, task)
+	}
+	// Wait for all PowerOff tasks to be finished. We intentionally ignore errors here
+	// because the VM may already got into PowerOff state and log the errors only.
+	// xref govc: https://github.com/vmware/govmomi/blob/512c168/govc/vm/destroy.go#L94-L96
+	if err := waitForTasksFinished(ctx, poweroffTasks, true); err != nil {
+		log.Info("Ignoring error for PowerOff task", "err", err)
+	}
+
+	destroyTasks := []*object.Task{}
+	for _, vm := range append(vmsToDeleteAndPoweroff, vmsToDelete...) {
+		log.Info("Destroying vm in vSphere", "vm", vm.managedObject.Config.Name)
+		if dryRun {
+			// Skipping actual destroy on dryRun.
+			continue
+		}
+		task, err := vm.object.Destroy(ctx)
+		if err != nil {
+			return err
+		}
+		destroyTasks = append(destroyTasks, task)
+	}
+	// Wait for all destroy tasks to succeed.
+	if err := waitForTasksFinished(ctx, destroyTasks, false); err != nil {
+		return errors.Wrap(err, "failed to wait for vm poweroff task to finish")
+	}
+
+	return nil
+}
+
+func waitForTasksFinished(ctx context.Context, tasks []*object.Task, ignoreErrors bool) error {
+	for _, t := range tasks {
+		if err := t.Wait(ctx); !ignoreErrors && err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// +kubebuilder:rbac:groups=ipam.cluster.x-k8s.io,resources=ipaddressclaims,verbs=get;list;watch;delete
+
+func (s *janitor) ipAddressClaims(ctx context.Context) error {
+	log := ctrl.LoggerFrom(ctx).WithName("IPAddressClaims")
+	ctrl.LoggerInto(ctx, log)
+	log.Info("Deleting IPAddressClaims")
+
+	// List all existing IPAddressClaims
+	ipAddressClaims := &ipamv1.IPAddressClaimList{}
+	if err := s.ipamClient.List(ctx, ipAddressClaims,
+		client.InNamespace(s.ipamNamespace),
+	); err != nil {
+		return err
+	}
+
+	errList := []error{}
+
+	for _, ipAddressClaim := range ipAddressClaims.Items {
+		ipAddressClaim := ipAddressClaim
+		// Skip IPAddressClaims which got created after maxCreationDate.
+		if ipAddressClaim.CreationTimestamp.After(s.maxCreationDate) {
+			continue
+		}
+
+		log.Info("Deleting IPAddressClaim", "IPAddressClaim", klog.KObj(&ipAddressClaim))
+
+		if s.dryRun {
+			// Skipping actual deletion on dryRun.
+			continue
+		}
+
+		if err := s.ipamClient.Delete(ctx, &ipAddressClaim); err != nil {
+			errList = append(errList, err)
+		}
+	}
+
+	return kerrors.NewAggregate(errList)
+}
+
+func (s *janitor) vSphereClusterModules(ctx context.Context) error {
+	log := ctrl.LoggerFrom(ctx).WithName("vSphere cluster modules")
```

Comment:
Is there any way we can consider maxCreationDate here?

I'm wondering if this code here could theoretically delete cluster modules an e2e test just created

Reply:
Okay then let's just see if we run into weird flakes